examples/sequential_retrieval.py

```diff
@@ -0,0 +1,422 @@
+"""
+# Recommending movies: retrieval using a sequential model
+
+In this example, we are going to build a sequential retrieval model. Sequential
+recommendation is a popular model that looks at a sequence of items that users
+have interacted with previously and then predicts the next item. Here, the order
+of the items within each sequence matters. So, we are going to use a recurrent
+neural network to model the sequential relationship. For more details,
+please refer to the [GRU4Rec](https://arxiv.org/abs/1511.06939) paper.
+
+Let's begin by importing all the necessary libraries, and setting the
+random seed for reproducibility.
+"""
+
+import os
+
+os.environ["KERAS_BACKEND"] = "jax"
+
+import collections
+import random
+
+import keras
+import pandas as pd
+import tensorflow as tf  # Needed only for the dataset
+from keras import ops
+
+import keras_rs
+
+keras.utils.set_random_seed(42)
+
+"""
+Let's also define all important variables/hyperparameters below.
+"""
+
+DATA_DIR = "./raw/data/"
+
+# MovieLens-specific variables
+MOVIELENS_1M_URL = "https://files.grouplens.org/datasets/movielens/ml-1m.zip"
+MOVIELENS_ZIP_HASH = (
+    "a6898adb50b9ca05aa231689da44c217cb524e7ebd39d264c56e2832f2c54e20"
+)
+
+RATINGS_FILE_NAME = "ratings.dat"
+MOVIES_FILE_NAME = "movies.dat"
+
+# Data processing args
+MAX_CONTEXT_LENGTH = 10
+MIN_SEQUENCE_LENGTH = 3
+TRAIN_DATA_FRACTION = 0.9
+
+RATINGS_DATA_COLUMNS = ["UserID", "MovieID", "Rating", "Timestamp"]
+MOVIES_DATA_COLUMNS = ["MovieID", "Title", "Genres"]
+MIN_RATING = 2
+
+# Training/model args
+BATCH_SIZE = 2048
+TEST_BATCH_SIZE = 2048
+EMBEDDING_DIM = 128
+NUM_EPOCHS = 10
+LEARNING_RATE = 0.05
+
+"""
+## Dataset
+
+Next, we need to prepare our dataset. Like we did in the
+[basic retrieval](https://github.com/keras-team/keras-rs/blob/main/examples/basic_retrieval.py)
+example, we are going to use the MovieLens dataset. 
+
+The dataset preparation step is fairly involved. The original ratings dataset
```

Comment:
Thanks for porting all the dataset massaging stuff here. That was most of the work I imagine.

Reply:
Yeah, pretty much. I was trying to not do the conversion to Pandas DF, and somehow use [`tf.data.group_by_window`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#group_by_window). At some point, I gave up xD